pkg/store/store.go

```diff
@@ -37,9 +38,45 @@ func NewAtomixStore(client atomix.Client) (Store, error) {
 	if err != nil {
 		return nil, err
 	}
-	return &atomixStore{
+	store := &atomixStore{
 		objects: objects,
-	}, nil
+		relations: relationMaps{
```

Comment:
I am not sure about that implementation for relation references. imo, If we end up adding internal maps and locks to the current data structure to achieve the goal then we are not really solving the problem if we don’t make it worse.  I expected to see references to the relations for each Entity rather than at store level (perhaps my understanding is wrong). Technically, you can have a list of incoming edges to the entity and outgoing edges from that entity but I am pretty sure there is a more elegant solution if we think about it a little bit more. I would suggest to consult with @kuujo   about the implementation details.

Reply:
Thank you, I was implementing it per Thomas' specifications, and I'll check with him and Jordan tomorrow.

Reply:
The idea was to keep an ephemeral, dynamic index of the references locally on each instance, entirely driven by events - hence able to respond to updates that happen elsewhere. The index can be maintained without regard to multi-node concurrency; only in-process thread concurrency is required. This is far more light-weight and since all write-locks happen in the event handler should avoid dead-locks. We only need read-locks in the get/list to populate the relation references - mostly as a courtesy to the apps - and also in the delete for consulting list of relations that should be removed before removing the entity. Maintaining the references in a distributed map would be a) expensive, b) complex and c) prone to dead-locks.